README
match  v0.16  -  by Michael W. Richmond <mwrsps@rit.edu>
(C)opyright 2000, 2001, 2002, 2003, 2007, 2010, 2011, 2012 Michael W. Richmond, all rights 
reserved

This package contains software to match one list of objects against
another list, allowing for arbitrary translation, rotation, scaling,
and some distortion.  The code is designed specifically for
astronomical data: list of stars or galaxies.  There is a small 
"pre-processor" to put astronomical data into the proper format 
for matching, and a "post-processor" to apply a transformation
to the astronomical coordinates in a list.  Most of the real work
is done by the "match" program itself.

The idea behind the matching of lists is to form a set of triangles
from the points in each list, and look for similar triangles.
The algorithm used in version 0.15 and 0.16 is based upon the article
"Fast Algorithms for Matching CCD Images to a Stellar Catalogue"
by Vello Tabur, Publications of the Astronomical Society of Australia,
vol 24, page 189 (2007).  See the following URL (which should
all be entered on one line)

     http://adsabs.harvard.edu/abs/2007PASA...24..189T


INSTALLATION

Follow the standard installation procedure:

  - uncompress the gzip'ed tar archive
  - cd match-0.16
  - ./configure
  - make

If your system has Perl installed, you can run a quick self-test:

  - make check

The test assumes that Perl is located in /usr/bin/perl.  If it's 
somewhere else, simply type

  - perl selftest.pl
    
See the documentation in "match.html" for a description of the 
programs, the input and output formats, etc.


REQUIREMENTS

The software is written in ANSI C, and requires nothing beyond the 
standard math library.  The selftest script is written in Perl;
it works with Perl 5, but fails with Perl 4.


NOTES

The pre- and post-processing steps assume that celestial coordinates
are expressed in decimal degrees, not sexigesimal notation.
That is,

        this:   233.0082           not this:  15:32:01.9

Incorporating support for sexigesimal notation is probably a Good
Idea for some future release.





NEWS
version 0.16  10/30/2012

    Small but important fix: modified the control in "calc_trans()"
    inside atpmatch.c so that if one tentative match fails, the
    code does NOT immediately quit; now, it will go on to other
    possible matches.  More robust, but also can now report
    less-than-optimal matches.  May require more tuning to find the
    optimal match.


version 0.15  08/17/2012

    Big change, removing the brute-force method of Valdes et al. 
    (PASP 107, 1119, 1995) for the more clever and robust technique
    of Tabur (PASJ 24, 189, 2007).  Tried to keep input/output/arguments
    the same, with very minor exceptions to input args.

	
version 0.14  11/04/2011

    Added "rotangle=" and "rottol=" arguments, so that the user can 
    now specify a permitted range of rotation angles between the
    two coordinate systems.

version 0.13  11/30/2010

    Added a Unix-style man page which is extracted by hand (ugh) from
    the .html manual.

    Added an "install" target for Makefile

version 0.12  11/22/2010

    Fixed the description of the "outfile=" argument in the USAGE section
    of "match.html".
   
    Added a printf statement in "make_vote_matrix()" in atpmatch.c.

version 0.11  6/26/2010

    Changed output TRANS elements format from %15.9f to %15.9e; should
    increase accuracy of subsequent calculations when values are small.

version 0.10  6/11/2008

    Small bug fix in "apply_match.c"

version 0.9   2/17/2007

    Added notes in documentation to warn against very large ranges
    in the input coordinate systems.  If the uncertainty in position
    of each item is more than about 50 units, matching will fail.
    See #define AT_MATCH_MAXDIST in atpmatch.h

version 0.8  10/25/2003

    Added an option to the "project_coords" program at the request of
    John Blakeslee and the ACS team.  The new "asec" option causes
    the program to print the (xi, eta) coords of each star in arcsec
    away from the projection point, rather than in radians.


version 0.7  5/19/2003

    Small fix to "apply_match" program: now handle coordinates near the
    celestial poles properly.
     -- MWR 5/19/2003


    Small bug fixes:
        a. match.c      if "reset_A_coords()" passed empty lists,
                                it no longer fails assertion, but prints
                                warning message and continues
        b. atpmatch.c   added similar checks in "atCalcRMS" and "atFindMedtf"
                                to deal sensibly with lists of zero stars
                                instead of failing assertions.
     -- MWR 10/23/2002


    Fixed "misc.c" so that "read_data_file" can now really handle data files
    with up to 20 columns; a bug prevented any data in columns >= 10 from
    being read properly.
     -- MWR 8/17/2002



version 0.6  6/19/2002

    Enhancement, code supplied by John Blakeslee
             new routine, atCalcRMS, calculates "sx" and "sy" for
             all matching items in each direction, using 1 iteration
             of 3-sigma clipping.  Adds two new fields to end of 
             the TRANS output line


version 0.5  12/31/2001

    Enhancements (thanks to John Blakeslee)
	      a. 'identity' option forces match with no scale change or rotation,
            with no translation (by default), or a given translation
            (if 'xsh=' and 'ysh=' arguments given)
         b. 'intrans' option allows user to specify exactly an initial
            TRANS structure
         c. after initial TRANS is discovered (or given by user), the
            program will always run 'atRecalcTrans' using only the 
            matching items

    Fixed several small bugs; thanks to John Blakeslee and Andrew
    Bennett.

    Added more test cases to 'selftest.pl'.


version 0.4  12/14/2001

    Fixed small bug in "project_coords.c" which limited number of
	 columns in data files to 10 (instead of the stated 20).

	 Modified "write_array" in atpmatch.c so that output files
	 "matched.mtA", "matched.mtB", "matched.unA", "matched.unB"
	 have coordinates printed with format 13.7f, instead of 12.6f.
	 When coordinates are radians, this provides granularity of
	 0.02 arcsec, versus 0.2 arcsec with old format.  The extra
	 precision may occasionally be warranted.

	 Added new elements to the TRANS structure, which provide a summary
    of the quality of the fit.  

    New option to produce a second line of output, the MEDTF structure,
    in case the two files are known to differ only by a pure translation.

    New "min_scale" and "max_scale" arguments are permitted, as well as
    the old "scale" argument.  



version 0.3   1/21/2001

    Added new command-line arguments "max_iter" and "halt_sigma",
    which allow user to control the number of iterations of
    the main program loop.  This allows the user to terminate
    the search for solutions when the residuals drop to a particular
    level, for example.

    Fixed small bug in "selftest.pl" script which caused the test
    of "id1" and "id2" options to fail.    MWR 10/29/2000

version 0.2   7/30/2000

    Added "id1=" and "id2=" options so that user-assigned ID numbers may
    optionally be used throughout program and kept in output files.


version 0.1   6/14/2000

    Initial release of the package.  Includes the workhorse program
    "match", plus a pre-processor (project_coords) and a post-processor
    (apply_match).  

    It would be nice to modify the pre-processor to handle (RA, Dec)
    coordinates expressed either in decimal degrees (in current version),
    or in sexigesimal notation.

    The regression test "selftest.pl" tests only the "match" program.
    There are no regression tests for "project_coords" or "apply_match".





misc.h
#if !defined(MISC_H)
#define MISC_H

/*
 *
 * FILE: misc.h
 * 
 * DESCRIPTION:
 * Definitions for structures and functions used in support
 * of the matching code.  Much of this is material that appears
 * in SHIVA, the SDSS environment.
 *
 */


#define SH_SUCCESS        0          /* indicates that all went well */
#define SH_GENERIC_ERROR  1          /* indicates that error occurred */


	/* a buffer used for parsing command-line arguments */
#define CMDBUFLEN       500

   /* max length of lines in input files */
#define LINELEN         300          

   /* ignore any lines in input files that start with this */
#define COMMENT_CHAR   '#'     

   /* data files can have this many data columns, at most */
#define MAX_DATA_COL    20

   /* each column in the data file can have at most this many characters */
#define MAX_COL_LENGTH 50


   /* possibilities for the "order" field of TRANS structures */
#define AT_TRANS_LINEAR      1           /* linear terms only */
#define AT_TRANS_QUADRATIC   2           /* linear plus quadratic */
#define AT_TRANS_CUBIC       3           /* linear plus quadratic plus cubic */

	/* maximum possible number of coefficients in a TRANS */
#define AT_TRANS_MAXCOEFF   16           /* for cubic case */


   /*
    * little wrappers around 'malloc' and 'free' functions.
    */

void *
shMalloc(int nbytes);

void
shFree(void *vptr);

void
shError(char *format, ...);

void
shFatal(char *format, ...);

void
shDebugSet(int level);

void
shDebug(int level, char *format, ...);

   /*
    * This is a preprocessor macro, which acts like a function.  The user
    * calls it with one argument, a condition which should evaluate to
    * 0 or 1.  If it evaluates to 1, then nothing happens; but if it
    * evaluates to 0, then the program prints an error message, giving
    * location of the error, and halts execution with an error code.
    * 
    * Thus, one typically uses it to make a 'sanity check', such as
    * making sure that a pointer (which really, really should have
    * a valid value) isn't NULL:
    *
    *     fp = fopen("file", "r");
    *        ...
    *     shAssert(fp != NULL);
    *     fgets(line, LINELEN, fp);
    *        ...
    */
#define shAssert(x) if((x)!=1){fprintf(stderr,"assertion fails in file %s, line %d\n",__FILE__,__LINE__);exit(1);}
 

   /*
    * This is a generic transformation from one coordinate system to another.
    * Given the measured (x, y), the transformed coords (x', y') are
    * calculated like this:
    *
    *   if linear terms only:
    *
    *       x' = A + B*x + C*y
    *       y' = D + E*x + F*y
    *
    *   if linear plus quadratic terms, 
    *
    *      x' =  A + Bx + Cy + Dxx + Exy + Fyy
    *      y' =  G + Hx + Iy + Jxx + Kxy + Lyy
    *
    *   if linear plus quadratic plus cubic, 
    *
    *      x' =  A + Bx + Cy + Dxx + Exy + Fyy + Gx(xx+yy) + Hy(xx+yy)
    *      y' =  I + Jx + Ky + Lxx + Mxy + Nyy + Ox(xx+yy) + Py(xx+yy)
    *
    *
    *  The 'order' field of the TRANS structure signals which
    *  of the above cases is to be used.
	 *
	 *  The 'nr' field contains the number of pairs ultimately used to 
	 *  determine the transform between the coordinate system.
	 *
	 *  The 'nm' field contains the number of pairs which -- after 
	 *  the transform has been determined and applied -- match up,
	 *  and can be used to determine the quality of the fit.
	 *
	 *  The 'sig' field holds the standard deviation of the separation
	 *  between matched stars in the two sets, after they have been
	 *  transformed into the coordinate system of the second list.
	 *  Only items which were used to derive the TRANS are included
	 *  in this calculation.
	 *
	 *  The 'sx' and 'sy' fields hold the standard deviation of the
	 *  1-D separations between corresponding items in the two sets,
	 *  after they have been transformed into the coordinate system
	 *  of the second list.  A single iteration of 3-sigma clipping
	 *  is used in the calculation.  
	 *  If the user specifies the "recalc" option, then the "sx" and "sy"
	 *  values will be based on all items which can be matched in 
	 *  the two frames -- not just those few which were used to derive
	 *  the TRANS.
    */

typedef struct Trans {
  int id;
  int order;
  double a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p;
  int nr;
  int nm;
  double sig;
  double sx;
  double sy;
} TRANS;

void atTransOrderSet(int order);
int atTransOrderGet(void);
TRANS *atTransNew(void);
TRANS *getGuessTrans(char *intransfile);
TRANS *getIdentityTrans(void);
void atTransDel(TRANS *trans);
void print_trans(TRANS *trans);
void copyTrans(TRANS *from, TRANS *to);

	/*
	 * The MEDTF structure holds statistics on the difference between
	 * matched pairs of stars, assuming that they can be matched
	 * by a simple translation -- without change in scale or rotation.
	 * Its elements are set by the atFindMedtf function, which is
	 * called only if the user specifies a command-line argument.
	 */

typedef struct Medtf {
	double mdx;          /* median offset in x-coord between lists */
	double mdy;          /* median offset in y-coord between lists */
	double adx;          /* average offset in x-coord between lists */
	double ady;          /* average offset in y-coord between lists */
	double sdx;          /* clipped stdev of differences in x-coord */
	double sdy;          /* clipped stdev of differences in y-coord */
	int nm;              /* number of matched pairs used in calculations */
} MEDTF;

MEDTF *atMedtfNew(void);
void atMedtfDel(MEDTF *medtf);
void print_medtf(MEDTF *medtf);


   /*
    * create a new s_star structure
    */

struct s_star * 
atStarNew(double x, double y, double mag);

   /*
    * read an ASCII file with a catalog of stars, and create a list
    * of s_star structures
    */
int
read_star_file(char *filename, int xcolumn, int ycolumn, int magcolumn,
               int idcolumn, int ra_hours_col, int *num_stars, 
               struct s_star **list);

   /*
    * read an ASCII file with a list of matched stars, created by
    * the atMatchLists function, and create a list of s_star structures.
    */
int
read_matched_file(char *filename, int *num_stars, struct s_star **list);

   /*
    * little routines to support read_star_file
    */
int
is_blank(char *line);

int
get_value(char *str, double *val);

#endif    /* MISC_H */



apply_match.c

/*
 *  match: a package to match lists of stars (or other items)
 *  Copyright (C) 2000  Michael William Richmond
 *
 *  Contact: Michael William Richmond
 *           Physics Department
 *           Rochester Institute of Technology
 *           85 Lomb Memorial Drive
 *           Rochester, NY  14623-5603
 *           E-mail: mwrsps@rit.edu
 *
 *  
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *  
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 */

/*
 * <AUTO>
 * FILE: apply_match.c
 *
 * <HTML>
 * Given 
 *    - an ASCII file consisting of a list of stars, with one line per
 *        star and multiple columns of information separated by white space
 *    - the numbers of the columns containing "X" and "Y" coords of stars
 *    - a central RA and Dec, each in decimal degrees
 *    - coefficients of a TRANS structure which convert "X" and "Y"
 *        coordinates from the ASCII file's system to plate coordinates
 *        (xi, eta), which are projections onto the tangent plane 
 *        centered on the central RA and Dec.
 *
 * run through the data file.  For each entry, calculate the (RA, Dec) of
 * the star, then replace the "X" and "Y" values with (RA, Dec).  Leave all
 * other information in the ASCII file as-is.
 *
 * The TRANS structure converts "X" and "Y" to (xi, eta) in one of three
 * ways.  If the user specifies a 'linear' transformation (which is the
 * default), then
 *
 *     xi = A + Bx + Cy
 *    eta = D + Ex + Fy
 *
 * In the case of 'quadratic', 
 *
 *     xi =  A + Bx + Cy + Dxx + Exy + Fyy
 *    eta =  G + Hx + Iy + Jxx + Kxy + Lyy
 *
 * In the case of 'cubic', 
 *
 *     xi =  A + Bx + Cy + Dxx + Exy + Fyy + Gx(xx+yy) + Hy(xx+yy)
 *    eta =  I + Jx + Ky + Lxx + Mxy + Nyy + Ox(xx+yy) + Py(xx+yy)
 *
 * where "xi" and "eta" are in radians, and measure the distance
 * of the star from the center of field from which the TRANS was calculated.
 * We assume that the given "ra" and "dec" values are the same as this
 * central position.
 *
 * We force all values of RA to lie between 0 < RA < 360
 *
 * Print the results to stdout, or place them into the file given
 * by the optional "outfile" command-line argument.
 *
 * Usage: apply_match starfile1 xcol ycol ra dec linear|quadratic|cubic
 *                    a b c d e f [g h i j k [l m n o ]] [outfile=] 
 *
 * </HTML>
 * </AUTO>
 *
 * modified to assume that the TRANS structure takes (x, y) coords and
 *   turns them into (xi, eta), in radians, rather than in arcseconds.
 *   MWR 5/24/2000
 *
 * modified to handle the three cases of linear, quadratic, or cubic
 *   TRANSformations. 
 *   MWR 6/11/2000
 *
 * fixed equations in proc_star_file() so that they handle properly
 *   the coordinate transformations near the celestial poles.
 *   MWR 5/19/2003
 *
 * added 10 more %s in the "sscanf" statement in proc_star_file()
 *   MWR 6/11/2008
 */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "misc.h"
#include "degtorad.h"

#undef DEBUG           /* get some of diagnostic output */


static int
proc_star_file(char *file, int xcol, int ycol, double ra, double dec,
               TRANS *trans, FILE *out_fp);


#define USAGE  "usage: apply_match starfile1 xcol ycol ra dec "\
                      "linear|quadratic|cubic "\
                      "a b c d e f [g h i j k [l m n o]] "\
                      "[outfile=] "
char *progname = "apply_match";



int
main
   (
   int argc,
   char *argv[]
   )
{ 
   char *fileA;
   int i;
   int xcol, ycol;
   int trans_order = AT_TRANS_LINEAR;
   int last_coeff;
   double ra, dec;
   double a, b, c, d, e, f, g, h, trans_i, j, k, l, m, n, o, p;
   char outfile[100];
   FILE *fp;
   TRANS trans;

   /* make special search for --version, --help, and help */
   if ((argc < 2) ||
       (strncmp(argv[1], "--help", 6) == 0) || 
       (strncmp(argv[1], "help", 4) == 0)) {
      printf("%s\n", USAGE);
      exit(0);
   }
   if (strncmp(argv[1], "--version", 9) == 0) {
      printf("%s: version %s\n", progname, VERSION);
      exit(0);
   }

   if (argc < 12) {
      fprintf(stderr, "%s\n", USAGE);
      exit(1);
   }

   fp = stdout;

   /* parse the arguments */
   fileA = argv[1];
   if (sscanf(argv[2], "%d", &xcol) != 1) {
      shFatal("invalid argument for column for X values in first file");
   }
   if (sscanf(argv[3], "%d", &ycol) != 1) {
      shFatal("invalid argument for column for Y values in first file");
   }
   if (sscanf(argv[4], "%lf", &ra) != 1) {
      shFatal("invalid argument for RA");
   }
   if (sscanf(argv[5], "%lf", &dec) != 1) {
      shFatal("invalid argument for Dec");
   }
   if (strncmp(argv[6], "linear", 6) == 0) {
      trans_order = AT_TRANS_LINEAR;
   }
   else if (strncmp(argv[6], "quadratic", 9) == 0) {
      trans_order = AT_TRANS_QUADRATIC;
   }
   else if (strncmp(argv[6], "cubic", 5) == 0) {
      trans_order = AT_TRANS_CUBIC;
   }
   else {
      shFatal("%s: invalid order keyword %s \n", argv[6]);
   }

   if (sscanf(argv[7], "%lf", &a) != 1) {
      shFatal("invalid argument for TRANS coefficient A");
   }
   if (sscanf(argv[8], "%lf", &b) != 1) {
      shFatal("invalid argument for TRANS coefficient B");
   }
   if (sscanf(argv[9], "%lf", &c) != 1) {
      shFatal("invalid argument for TRANS coefficient C");
   }
   if (sscanf(argv[10], "%lf", &d) != 1) {
      shFatal("invalid argument for TRANS coefficient D");
   }
   if (sscanf(argv[11], "%lf", &e) != 1) {
      shFatal("invalid argument for TRANS coefficient E");
   }
   if (sscanf(argv[12], "%lf", &f) != 1) {
      shFatal("invalid argument for TRANS coefficient F");
   }
   last_coeff = 12;

   if ((trans_order == AT_TRANS_QUADRATIC) || (trans_order == AT_TRANS_CUBIC)) {
      if (sscanf(argv[13], "%lf", &g) != 1) {
         shFatal("invalid argument for TRANS coefficient G");
      }
      if (sscanf(argv[14], "%lf", &h) != 1) {
         shFatal("invalid argument for TRANS coefficient H");
      }
      if (sscanf(argv[15], "%lf", &trans_i) != 1) {
         shFatal("invalid argument for TRANS coefficient I");
      }
      if (sscanf(argv[16], "%lf", &j) != 1) {
         shFatal("invalid argument for TRANS coefficient J");
      }
      if (sscanf(argv[17], "%lf", &k) != 1) {
         shFatal("invalid argument for TRANS coefficient K");
      }
      if (sscanf(argv[18], "%lf", &l) != 1) {
         shFatal("invalid argument for TRANS coefficient L");
      }
      last_coeff = 18;
   }
   if (trans_order == AT_TRANS_CUBIC) {
      if (sscanf(argv[19], "%lf", &m) != 1) {
         shFatal("invalid argument for TRANS coefficient M");
      }
      if (sscanf(argv[20], "%lf", &n) != 1) {
         shFatal("invalid argument for TRANS coefficient N");
      }
      if (sscanf(argv[21], "%lf", &o) != 1) {
         shFatal("invalid argument for TRANS coefficient O");
      }
      if (sscanf(argv[22], "%lf", &p) != 1) {
         shFatal("invalid argument for TRANS coefficient P");
      }
      last_coeff = 22;
   }

   /* 
    * check for optional argument "outfile"
    */
   for (i = last_coeff + 1; i < argc; i++) {
      if (strncmp(argv[i], "outfile=", 8) == 0) {
         if (sscanf(argv[i] + 8, "%s", outfile) != 1) {
            shFatal("invalid argument for outfile argument");
         }
      }
      if ((fp = fopen(outfile, "w")) == NULL) {
         shFatal("can't open file %s for output", outfile);
      }
   }

   /* create a TRANS structure from the given coefficients */
   trans.order = trans_order;
   trans.a = a;
   trans.b = b;
   trans.c = c;
   trans.d = d;
   trans.e = e;
   trans.f = f;
   trans.g = g;
   trans.h = h;
   trans.i = trans_i;
   trans.j = j;
   trans.k = k;
   trans.l = l;
   trans.m = m;
   trans.n = n;
   trans.o = o; 
   trans.p = p;

   /* now walk through the file and do the dirty work */
   if (proc_star_file(fileA, xcol, ycol, ra, dec, &trans, fp) != SH_SUCCESS) {
      shError("can't process data from file %s", fileA);
      exit(1);
   }


   if (fp != stdout) {
      fclose(fp);
   }

   return(0);
}



/****************************************************************************
 * ROUTINE: proc_star_file
 *
 * walk through the given file, one line at a time.  
 *
 * If the line starts with COMMENT_CHAR, place it into the output stream.
 * If the line is completely blank, place it into the output stream.
 *
 * Otherwise, 
 *   - read in the entire line, 
 *   - figure out the "X" and "Y" coords
 *   - transform the "X" and "Y" coords to be (RA, Dec) from the central
 *         "ra" and "dec", in units of arcseconds
 *   - transform from the tangent plane back to the spherical sky, so that
 *         we have genuine (RA, Dec) for each star
 *   - print out the line, replacing the "X" with RA, the "Y" with Dec
 *
 * RETURNS:
 *   SH_SUCCESS            if all goes well
 *   SH_GENERIC_ERROR      if not
 */

static int
proc_star_file
   (
   char *file,              /* I: name of input file with star list */
   int xcol,                /* I: position of column with X positions */
   int ycol,                /* I: position of column with Y positions */
   double ra,               /* I: central RA of tangent plane (degrees) */
   double dec,              /* I: central Dec of tangent plane (degrees) */
   TRANS *trans,            /* I: TRANS taking (x,y) -> (ra, dec) */
   FILE *out_fp             /* I: place output into this stream */
   )
{
   char line[LINELEN];
   char col[MAX_DATA_COL + 1][MAX_COL_LENGTH + 1];
   int i, ncol;
   int last_column = -1;
   double xval, yval;
   double r_ra, r_dec;
   double z, alpha, delta;
   double delta_ra, delta_dec;
   double rsquared;
   FILE *in_fp;

   if ((in_fp = fopen(file, "r")) == NULL) {
      shError("proc_star_file: can't open file %s for input", file);
      return(SH_GENERIC_ERROR);
   }

   last_column = (xcol > ycol ? xcol : ycol);
   r_ra = ra*DEGTORAD;
   r_dec = dec*DEGTORAD;

   while (fgets(line, LINELEN, in_fp) != NULL) {

      if (line[0] == COMMENT_CHAR) {
        continue;
      }
      if (is_blank(line)) {
        continue;
      }

      shAssert(MAX_DATA_COL >= 20);
      ncol = sscanf(line, "%s %s %s %s %s %s %s %s %s %s %s %s %s %s %s %s %s %s %s %s", 
              &(col[0][0]), &(col[1][0]), &(col[2][0]), &(col[3][0]), 
              &(col[4][0]), &(col[5][0]), &(col[6][0]), &(col[7][0]), 
              &(col[8][0]), &(col[9][0]), 
              &(col[10][0]), &(col[11][0]), &(col[12][0]), &(col[13][0]),
              &(col[14][0]), &(col[15][0]), &(col[16][0]), &(col[17][0]),
              &(col[18][0]), &(col[19][0]));
      if (last_column > ncol) {
         shError("proc_star_file: not enough entries in following line; skipping");
         shError("  %s", line);
         continue;
      }
        
      /* now read values from each column */
      if (get_value(col[xcol], &xval) != SH_SUCCESS) {
         shError("read_data_file: can't read X value from %s; skipping", 
                  col[xcol]);
         continue;
      }
      if (get_value(col[ycol], &yval) != SH_SUCCESS) {
         shError("read_data_file: can't read Y value from %s; skipping", 
                  col[ycol]);
         continue;
      }


      /* 
       * let's transform from (x,y) to (delta_ra, delta_dec),
       * using either a linear, quadratic, or cubic transformation
       * (signalled by the 'order' field of the TRANS)
       */
      switch (trans->order) {
      case AT_TRANS_LINEAR:
         delta_ra  = trans->a + trans->b*xval + trans->c*yval;
         delta_dec = trans->d + trans->e*xval + trans->f*yval;
         break;
      case AT_TRANS_QUADRATIC:
         delta_ra  = trans->a + trans->b*xval + trans->c*yval +
                        trans->d*xval*xval + trans->e*xval*yval + 
                        trans->f*yval*yval;
         delta_dec = trans->g + trans->h*xval + trans->i*yval +
                        trans->j*xval*xval + trans->k*xval*yval + 
                        trans->l*yval*yval;
         break;
      case AT_TRANS_CUBIC:
         rsquared  = xval*xval + yval*yval;
         delta_ra  = trans->a + trans->b*xval + trans->c*yval +
                        trans->d*xval*xval + trans->e*xval*yval + 
                        trans->f*yval*yval +
                           trans->g*xval*rsquared + trans->h*yval*rsquared;
         delta_dec = trans->i + trans->j*xval + trans->k*yval +
                        trans->l*xval*xval + trans->m*xval*yval + 
                        trans->n*yval*yval +
                           trans->o*xval*rsquared + trans->p*yval*rsquared;
         break;
      default: 
         shFatal("%s: proc_star_file: invalid trans->order %d", 
                    progname, trans->order);
         break;
      }
      

#if 0
      /*
       * and now convert from arcseconds to radians 
       * (convenient for calculations) 
       */
      delta_ra = (delta_ra/3600.0)*DEGTORAD;
      delta_dec = (delta_dec/3600.0)*DEGTORAD;
#endif

      /* 
       * we have (delta_ra, delta_dec), in radians; these give the distance
       * of this star from the central (RA, Dec).  Now we can de-project from
       * the tangent plane (centered on RA,Dec) and calculate the actual
       * RA, Dec of the star (in degrees)
       */
	{ 
	double zz;

		z = cos(r_dec) - delta_dec*sin(r_dec);
		zz = atan2(delta_ra, z)/DEGTORAD;
		alpha = zz + ra;

      zz = cos((alpha - ra)*DEGTORAD)*(sin(r_dec) + delta_dec*cos(r_dec));
		delta = atan2(zz, z)/DEGTORAD;
	}


      /*
       * make sure new RA lies in range  0 < RA < 360
       */
      if (alpha < 0) {
        alpha += 360.0;
      } 
      if (alpha >= 360.0) {
        alpha -= 360.0;
      }

		/* 
		 * make sure Dec lies in range  -90 < Dec < +90
		 */
		if (delta < -90) {
			delta += 180;
		}
		if (delta > 90) {
			delta -= 180;
		}

#ifdef DEBUG
      printf(" new RA = %10.5f,   new dec = %10.5f\n", alpha, delta);
#endif
      
      /* now build up the output line */
      line[0] = '\0';
      for (i = 0; i < ncol; i++) {
        if (i == xcol) {
          sprintf(line, "%s %10.5f", line, alpha);
        }
        else if (i == ycol) {
          sprintf(line, "%s %10.5f", line, delta);
        } 
        else {
          sprintf(line, "%s %s", line, col[i]);
        }
      }
      fprintf(out_fp, "%s\n", line);

   }



   fclose(in_fp);
   return(SH_SUCCESS);
}
